src/facade/dragonfly_connection.cc

```diff
@@ -2173,22 +2177,28 @@ void Connection::SetPipelineLowBoundStats(unsigned limit) {
 
 Connection::WeakRef::WeakRef(std::shared_ptr<Connection> ptr, unsigned thread_id,
                              uint32_t client_id)
-    : ptr_{std::move(ptr)}, thread_id_{thread_id}, client_id_{client_id} {
-}
-
-unsigned Connection::WeakRef::Thread() const {
-  return thread_id_;
+    : ptr_{std::move(ptr)}, last_known_thread_id_{thread_id}, client_id_{client_id} {
 }
 
 Connection* Connection::WeakRef::Get() const {
-  // We should never access the connection object from other threads.
-  DCHECK_EQ(ProactorBase::me()->GetPoolIndex(), int(thread_id_));
-
-  //  The connection can only be deleted on this thread, so
-  //  this pointer is valid until the next suspension.
-  //  Note: keeping a shared_ptr doesn't prolong the lifetime because
-  //  it doesn't manage the underlying connection. See definition of `self_`.
-  return ptr_.lock().get();
+  auto sptr = ptr_.lock();
+
+  if (!sptr) {
+    return nullptr;
+  }
+
+  // We potentially read a memory location that is written by another thread.
+  // It is fine because we know the location exists, so it is safe to read.
```

Comment:
Just because the location exists doesn't mean it's not problematic. Concurrent read/writes are a data race and data races according to the cpp standard slowly enter the UB realm.

IMO `OnPostMigrateThread` sets `owning_thread_id_` and it runs on the newly migrated thread. So:

thread a -> traverses the weak refs it just fetched, and is about to do read `owning_thread_id`
thread b -> we just migrated so we are running `OnPostMigrateThread` which sets the `owning_thread_id` 

That's a clear data race. 

I don't think the chances of this happening are big, but it's a possibility. Why don'y we make the `owning_thread_id` atomic if it can be potentially be accessed concurrently ?